src/caslist/mod.rs

```diff
@@ -0,0 +1,103 @@
+use std::marker::PhantomData;
+use std::iter::IntoIterator;
+use std::mem;
+use std::ptr;
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+#[cfg(test)]
+mod test;
+
+/// A simple linked list, based on compare-and-swap. Values can only
+/// be prepended to the front. At present, they can never be removed.
+pub struct CasList<T> {
+    head: AtomicPtr<Cell<T>>, // if not null, a unique, transmuted ptr to a Box<T>
+}
+
+struct Cell<T> {
+    next: *mut Cell<T>, // if not null, a unique, transmuted ptr to a Box<T>
+    data: T,
+}
+
+impl<T> CasList<T> {
+    pub fn new() -> Self {
+        CasList {
+            head: AtomicPtr::new(ptr::null_mut())
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        let head = self.head.load(Ordering::Acquire);
+        head.is_null()
+    }
+
+    pub fn prepend(&self, value: T) {
+        unsafe {
+            let cell = Box::new(Cell {
+                next: ptr::null_mut(),
+                data: value
+            });
+
+            // We're going to transfer ownership of this box into the
+            // list. Note that none of this code below can panic,
+            // which is important for avoiding leaks!
+            let cell: *mut Cell<T> = mem::transmute(cell);
+            loop {
+                let head = self.head.load(Ordering::Relaxed);
+                (*cell).next = head;
+                if self.head.compare_and_swap(head, cell, Ordering::Release) == head {
```

Comment:
You can make this loop faster under contention, like this:
```rust
let mut head = self.head.load(Ordering::Acquire);
loop {
    (*cell).next = head;
    let previous = self.head.compare_and_swap(head, cell, Ordering::AcqRel);
    if previous == head {
        break;
    } else {
        head = previous;
    }
}
```

Also, your orderings are incorrect - you need `Acquire` and `AcqRel`. :)
The way to think about this is the following...

Suppose Bob wants to prepend a new value to the list. He loads the current head with `Relaxed` ordering (like you did). Then he sets the next field of the new cell and installs a new cell with `Release` ordering.

Now Alice comes and loads the head with `Acquire` ordering. She gets value `h`, which is the head Bob just installed. Because she used `Acquire` ordering, she will see all writes to memory that happened before `h` was installed. This is good.

However, will she see writes to memory that happened before `(*h).next` was installed? Unfortunately, no, and that is because Bob used `Relaxed` load! Had he used `Acquire` load, then Alice would see even those writes.

With correct ordering, this is what would happen...
Think of this as a long synchronization chain. Alice's load (`Acquire`) synchronizes with Bob's store (`Release`), which happened before he wrote to `(*cell).next`, and that load (`Acquire`) synchronized with the previous store (`Release`) to the head (whoever did it) and so forth...

So the key take away is: make sure to chain these operations! `Relaxed` is often fishy. :)